### PR TITLE
Path params validation

### LINKED ISSUE #690: Path params are not validated
**Describe the bug**

Path-level path parameters are not validated with their schema.

**To Reproduce**

```yaml
openapi: 3.0.0
servers:
  - url: 'http://localhost:3000'
paths:
  '/path/{param}':
    parameters:
      - type: string
        name: param
        in: path
        required: true
        schema:
          type: string
          pattern: '^[A-Z]$'
    get:
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                type: object
                properties: {}
```

or (get-level)

```yaml
openapi: 3.0.0
servers:
  - url: 'http://localhost:3000'
paths:
  '/path/{param}':
    get:
      parameters:
        - type: string
          name: param
          in: path
          required: true
          schema:
            type: string
            pattern: '^[A-Z]$'
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                type: object
                properties: {}
```
```

```
curl 127.0.0.1:4010/path/123
```

**Expected behavior**

Get a validation error in response.

**Actual**

I get empty JSON object and OK status

**Environment (remove any that are not applicable):**
 - Library version: v3.1.1

### PR DESCRIPTION
**Pull request by StefanDywersant**
Closes #690 

## Checklist

- [x] Tests have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## What kind of change does this PR introduce?

Feature

## What is the current behavior? What is the new behavior?

Path parameters are now validated against schema.

## Does this PR introduce a breaking change?

It does if someone relies on path params is not validated.

### CODE REVIEW

File: test-harness/specs/parameters-ac9.oas3.txt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+Then I should get a 422 response back
+And the payload of the response should be in application/problem+json content-type
+====spec====
+openapi: 3.0.0
+paths:
+  '/path/{param}':
+    parameters:
+      - name: param
+        in: path
+        required: true
```
[the remark this answers is not in the document]
Comment (philsturgeon) [reply test-harness/specs/parameters-ac9.oas3.txt]: If a path param is not provided in this path `/foo/{bar}/baz` that would be `/foo//baz`? That would just be a 404 I guess?

[the remark this answers is not in the document]
Comment (XVincentX) [reply test-harness/specs/parameters-ac9.oas3.txt]: Good point, we can either take this as a validation error (missing required parameter) or return 404 as a missing route error. Hard to say for me which is the best.

File: packages/http/src/validator/deserializers/index.ts [hunk 2]
```diff
@@ -18,4 +20,10 @@ export const query = new HttpParamDeserializerRegistry([
   new DeepObjectStyleDeserializer(),
 ]);
 
+export const path = new HttpParamDeserializerRegistry([
+  new SimpleStyleDeserializer(),
+  new LabelStyleDeserializer(),
+  new MatrixStyleDeserializer(),
+]);
```
Comment (XVincentX) [inline packages/http/src/validator/deserializers/index.ts]: One thing to keep in mind is that we can in the future employ a library doing all this for us (https://github.com/litejs/uri-template-lite#readme) — this apparently passes all the URI Templates specification tests and has both a "match", and "expand" function we could use.

@StefanDywersant Would it be possible to do a quick (no more than 1H) test on how hard would it be to use this library instead of writing our deserialiser?

File: packages/http/src/validator/deserializers/style/__tests__/label.spec.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+import { LabelStyleDeserializer } from '../label';
+import * as createObjectFromKeyValListModule from '../utils';
+
+describe('LabelStyleDeserializer', () => {
+  const labelStyleDeserializer = new LabelStyleDeserializer();
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
```
Comment (XVincentX) [inline packages/http/src/validator/deserializers/style/__tests__/label.spec.ts]: Is this needed? We're not mocking anything in this test file.

File: packages/http/src/__tests__/fixtures/index.ts [hunk 4]
```diff
@@ -274,7 +274,7 @@ export const httpInputs: IHttpRequest[] = [
   {
     method: 'post',
     url: {
-      path: '/',
+      path: '/todos',
```
Comment (XVincentX) [inline packages/http/src/__tests__/fixtures/index.ts]: Why this change?

File: packages/http/src/validator/validators/__tests__/__snapshots__/query.spec.ts.snap [hunk 5]
```diff
@@ -1,28 +1,28 @@
...
-    "code": "required",
-    "message": "should have required property 'aparam'",
+    "code": "deprecated",
+    "message": "Query param param is deprecated",
     "path": Array [
       "query",
+      "param",
     ],
-    "severity": 0,
+    "severity": 1,
```
Comment (XVincentX) [inline packages/http/src/validator/validators/__tests__/__snapshots__/query.spec.ts.snap]: Aha ok it took me ages to understand what was going here — all it's happening is simply reversing the order :trollface:

Comment (StefanDywersant) [reply packages/http/src/validator/validators/__tests__/__snapshots__/query.spec.ts.snap]: Haha, me too ;) I should have comment it somehow to save reviewers time.

File: packages/http/src/validator/index.ts [hunk 6]
```diff
@@ -35,7 +37,8 @@ const validateInput: ValidatorFn<IHttpOperation, IHttpRequest> = ({ resource, el
 
   return results
     .concat(headersValidator.validate(element.headers || {}, (request && request.headers) || []))
-    .concat(queryValidator.validate(element.url.query || {}, (request && request.query) || []));
+    .concat(queryValidator.validate(element.url.query || {}, (request && request.query) || []))
+    .concat(pathValidator.validate(getPathParams(element.url.path, resource.path), (request && request.path) || []));
```
[the remark this answers is not in the document]
Comment (StefanDywersant) [reply packages/http/src/validator/index.ts]: It is exactly as you stated: `path` is a string. There is `getPathParams` function if you take a closer look.

[the remark this answers is not in the document]
Comment (StefanDywersant) [reply packages/http/src/validator/index.ts]: However, I'd vote on extracting path params at an earlier stage. They are needed here but will be needed also for WebHooks. How do you feel about that?

File: packages/http/src/validator/index.ts [hunk 7]
```diff
@@ -77,4 +80,14 @@ const validateOutput: ValidatorFn<IHttpOperation, IHttpResponse> = ({ resource,
   );
 };
 
+function getPathParams(path: string, template: string) {
+  const matches = new RegExp('^' + template.replace(/{(.+?)}/g, (_, name) => `(?<${name}>.*)`)).exec(path);
```
[the remark this answers is not in the document]
Comment (StefanDywersant) [reply packages/http/src/validator/index.ts]: Believe me or not but I did some lookup over the internet. I apparently failed.

File: packages/http/src/validator/__tests__/index.spec.ts [hunk 8]
```diff
@@ -122,23 +131,73 @@ describe('HttpValidator', () => {
...
 
       describe('input.url.query is not set', () => {
         it("validates query assuming it's empty", validate(undefined, { url: { path: '/' } }));
       });
     });
+
+    describe('path validation in enabled', () => {
+      describe('request is set', () => {
+        describe('request.path is set', () => {
+          it('validates path', () => {
```
[the remark this answers is not in the document]
Comment (StefanDywersant) [reply packages/http/src/validator/__tests__/index.spec.ts]: 👍

File: packages/http/src/validator/__tests__/index.spec.ts [hunk 9]
```diff
@@ -1,7 +1,15 @@
...
-import { bodyValidator, headersValidator, queryValidator, validateInput, validateOutput } from '../index';
+import {
+  bodyValidator,
+  headersValidator,
+  pathValidator,
+  queryValidator,
+  validateInput,
+  validateOutput
+} from '../index';
+import { HttpParamStyles } from '@stoplight/types/dist';
```
[the remark this answers is not in the document]
Comment (StefanDywersant) [reply packages/http/src/validator/__tests__/index.spec.ts]: Done 👍

File: packages/http/src/validator/deserializers/style/__tests__/label.spec.ts (hunk 3 above)
Comment (StefanDywersant) [reply packages/http/src/validator/deserializers/style/__tests__/label.spec.ts]: Definitely not, I also removed it from the rest of deserializer tests. Good catch & bit smaller codebase.

File: packages/http/src/__tests__/fixtures/index.ts (hunk 4 above)
Comment (StefanDywersant) [reply packages/http/src/__tests__/fixtures/index.ts]: Not sure, most probably some leftover. I removed it.

File: packages/http/src/validator/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (XVincentX) [reply packages/http/src/validator/index.ts]: Yeah so the problem right now is that we're kind of using two Uri Template libraries

`urijs` (which is being removed in #669)

[the remark this answers is not in the document]
Comment (XVincentX) [reply packages/http/src/validator/index.ts]: Yeah so the thing is we are already an Uri Template library `urijs` — that is being removed in #669;

Then we have another one `uri-templates` that is used for the example generation; we need to verify which library among the following ones can do the following things:

1. Parsing and expanding (regular case)
2. Desubstitution (grab parameters from a real url and its template)

Here are the libraries we have:

* https://medialize.github.io/URI.js/uri-template.html
* https://github.com/geraintluff/uri-templates
* https://github.com/litejs/uri-template-lite

File: test-harness/specs/parameters-ac9.oas3.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (StefanDywersant) [reply test-harness/specs/parameters-ac9.oas3.txt]: Prism returns 422 for `/foo//baz`. I just added a test case covering that.

1) 422 is a correct error for such case:
a) the path is known (thus not 404) however, the `bar` param is missing.
b) imagine that `bar` is defined as a number & request path = `/foo/2118/baz` - this also results in 422
c) having that said, invalid format and missing param are similar cases and should both lead to 422

2. On the other hand we have a case where param is last part of path: `/foo/{bar}`.
The input `/foo` results in 404 (test case: https://github.com/stoplightio/prism/blob/master/test-harness/specs/parameters-ac1.oas3.txt).

Even though 2) I think 1) is still correct behavior.

File: packages/http/src/validator/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (StefanDywersant) [reply packages/http/src/validator/index.ts]: The full var name is `resource.request.path`. These are out request path parameters specifications (confused with that "api desc" stuff, how should I name it here?), not actual request input values.

[the remark this answers is not in the document]
Comment (StefanDywersant) [reply packages/http/src/validator/index.ts]: Aha, and the source of confusion might be that we have:
`operation.url.path: string` - the actual path
`operation.request.path: stuff[]` - path param specs

File: packages/http/src/validator/deserializers/index.ts (hunk 2 above)
Comment (StefanDywersant) [reply packages/http/src/validator/deserializers/index.ts]: There are two things making it a little harder to implement:
1) OpenApi's param serialization is a superset of what is inside URI Template RFC (deepobject, form, ..)
2) in order to perform extraction we'd need to generate URI Template - again, OpenApi does not help us here because OAS's `path` is not a URI Template.

File: packages/http/src/validator/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (StefanDywersant) [reply packages/http/src/validator/index.ts]: I'd initially vote in the `-lite` lib, however it:
```
(karol@MacBook-Pro-Karol prism)$ yarn add -W --dev @types/uri-template-lite
yarn add v1.19.1
[1/4] 🔍  Resolving packages...
error An unexpected error occurred: "https://registry.yarnpkg.com/@types%2furi-template-lite: Not found".
```

File: packages/http/src/validator/deserializers/index.ts (hunk 2 above)
Comment (philsturgeon) [reply packages/http/src/validator/deserializers/index.ts]: yeah all the csv supporting, split, matrix, pipes, etc is all OpenAPI-specific. This is not a complication of OpenAPI, more of a recognitiion of all the things people do in the wild which do not perfectly rely on standards. If everything was perfectly built using defined standards, our tooling would be easier!

**Comment by philsturgeon** [comment]
Is there somewhere in this PR, or elsewhere in the existing codebase, where we deal with the awkard logic for "each type of parameter style has a different defaults for types? This stuff: 

image

File: test-harness/specs/parameters-ac9.oas3.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (philsturgeon) [reply test-harness/specs/parameters-ac9.oas3.txt]: It's probably just a personal opinion but I would expect `/foo//bar` to be a 404. Path parameters are a bit of a invented concept which doesnt really exist in HTTP, so... considering it an "invalid parameter" is mostly just down to us naming everything "validation" when really it is "looking for various types of failures are responding accordingly". 

What happens if you go to `https://github.com/stoplightio//pull/702`? 

For other failures, I think we could probably get some clarity here by checking common JavaScript routing libraries. Any routing system which allows a `/foo/:integer/bar`? If you send a string in as that param, does that 404 or 422? 

I think we consider too many things "validation", and that is confusing us when it comes to figuring out what prism should do.

**Comment by StefanDywersant** [comment]
> Is there somewhere in this PR, or elsewhere in the existing codebase, where we deal with the awkard logic for "each type of parameter style has a different defaults for types? This stuff:
> 
> image

Path (in this PR): https://github.com/stoplightio/prism/pull/702/files#diff-5c7aab71e5a0049cfb2006afe47c8a35R11
Query: https://github.com/stoplightio/prism/pull/702/files#diff-5c7aab71e5a0049cfb2006afe47c8a35R11
Headers: https://github.com/stoplightio/prism/blob/master/packages/http/src/validator/validators/headers.ts#L12
Body: might be missing - need to double check

File: packages/http/src/validator/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (XVincentX) [reply packages/http/src/validator/index.ts]: Yeah. We can leave as it is though, no actions point for this.

File: test-harness/specs/parameters-ac9.oas3.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (XVincentX) [reply test-harness/specs/parameters-ac9.oas3.txt]: @StefanDywersant How hard would it be to treat this as a 404?

File: packages/http/src/validator/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (XVincentX) [reply packages/http/src/validator/index.ts]: Don't be discouraged by that; we can easily provide the types for that and contribute back to the community. If the library quality is effectively better, let's go for it.

File: packages/http/src/validator/deserializers/index.ts (hunk 2 above)
Comment (XVincentX) [reply packages/http/src/validator/deserializers/index.ts]: What did we do to deserve this? :trollface: — resolving :)

File: packages/http/package.json [hunk 10]
```diff
@@ -32,6 +32,7 @@
     "pino": "^5.13.2",
     "tslib": "^1.10.0",
     "type-is": "^1.6.18",
+    "uri-templates": "^0.2.0",
```
Comment (XVincentX) [inline packages/http/package.json]: Can we use the same library everywhere? URIJS is going away with proxy command, but the example generation is using another library (see the CLI). It'd be cool to replace it and use the same here as well.

Comment (StefanDywersant) [reply packages/http/package.json]: My bad that I did not unified those libraries.

I looked on [uri-templates](https://github.com/geraintluff/uri-templates#readme) and [uri-template-lite](https://github.com/litejs/uri-template-lite#readme). Both seem to support full spec. The latter is 3x smaller and the code looks less dense. My vote goes on `-lite` variant.

Now, cli and http uses `uri-template-lite`.

File: packages/cli/src/util/__tests__/paths.spec.ts [hunk 11]
```diff
@@ -75,7 +75,7 @@ describe('createExamplePath()', () => {
           },
           responses: [{ code: '200' }],
         }),
-        r => expect(r).toEqual('/path?p[a][aa]=1&p[a][ab]=2'),
+        r => expect(r).toEqual('/path?p%5Ba%5D%5Baa%5D=1&p%5Ba%5D%5Bab%5D=2'),
```
Comment (StefanDywersant) [inline packages/cli/src/util/__tests__/paths.spec.ts]: This an similar lines need comment: the library is performing decodeURIComponent on passed arguments. That actually fixes a "bug". I used quotes because both versions put into the browser will work. The current one is more correct.

File: test-harness/specs/parameters-ac9.oas3.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (StefanDywersant) [reply test-harness/specs/parameters-ac9.oas3.txt]: The logic for checking `deprecated` and `required` is one place for all parameter types (query, body, header, and path). It would require some gymnastics to treat path-param-required differently. If I add that in this PR it will be soon redone during validators refactor. I'd do that inside validator-refactor PR to avoid doubling the work.

Additionally, I still have some doubts about whether it is a good approach:

`https://github.com/stoplightio//pull/702` - returns 404 because there is no such repo called ''. That's because it would not pass validation during repo creation.

Let's consider another example. Github repo-name-validation disallows '_'.

`https://github.com/stoplightio/_gigi_dagostino/pull/702` - shall it return 404 or 422?

I'm sure that it will respond with 404, again, the repo can't exist because it would not pass repo-name-validation. I'd say that 404 is a data-driven error and Prism has no database.


Summary:
`Prism pretends that it has some database and guesses whether a resource exists`
vs
`Prism focuses on request validation against the API description`

or

`Developers are lazy and don't validate input params if they don't need to, so Prism is`
vs
`Developers are lazy and don't validate input params if they don't need to, so Prism isn't`

File: packages/cli/src/util/paths.ts [hunk 12]
```diff
@@ -29,7 +29,7 @@ export function createExamplePath(operation: IHttpOperation): Either.Either<Erro
         }),
       );
     }),
-    Either.map(({ template, values }) => parse(template).expand(values)),
+    Either.map(({ template, values }) => new URI.Template(template).expand(values)),
```
Comment (XVincentX) [inline packages/cli/src/util/paths.ts]: We can avoid to create an instance and collect it right away.

```suggestion
    Either.map(({ template, values }) => URI.expand(template, values)),
```

File: packages/cli/src/util/__tests__/paths.spec.ts (hunk 11 above)
Comment (XVincentX) [reply packages/cli/src/util/__tests__/paths.spec.ts]: Hmm, I am afraid I do not understand what's going on here — can you elaborare a little bit more?

File: test-harness/specs/parameters-ac9.oas3.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (XVincentX) [reply test-harness/specs/parameters-ac9.oas3.txt]: Ok I _think_ I understand what you're saying.

The basic idea would be that since Prism has the knowledge of your API it can proactively say "hey I know what you're trying to do — and you're missing a parameter so here's a nice 422 saying that this is not ok" — while a regular web server won't effectively know what's going on and has no choice than answering 404.

Interesting; on one side, we can provide a helpful error message to the user; on the other, we want to try to behave as much as possible as a real web server.

> It would require some gymnastics to treat path-param-required differently

Yeah I understand; moreover this kind of error should be thrown by the router component and not the deserialiser/validator. It's kind of a big effort right now and I think it's be better to wait until the validation stuff has been refactored.

I vote to keep (at least temporarily) the 422 status code, and reason about this in the future

File: packages/cli/src/util/paths.ts (hunk 12 above)
Comment (StefanDywersant) [reply packages/cli/src/util/paths.ts]: Done 👍

File: packages/cli/src/util/__tests__/paths.spec.ts (hunk 11 above)
Comment (StefanDywersant) [reply packages/cli/src/util/__tests__/paths.spec.ts]: Sure. Previous library (`uri-templates`) did not care about whether input params are already uri encoded or not. That's why chars like '[' or ']' were visible in final urls. The new lib - `uri-template-lite` does encodeUriComponent on each params - that's the source of difference.

Both forms are valid though if you put it to the browser (browsers automatically encode some special characters) but I believe it is better that we provide already encoded form. So no relaying on browsers, our example uris will be ready to use.

File: test-harness/specs/parameters-ac9.oas3.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (StefanDywersant) [reply test-harness/specs/parameters-ac9.oas3.txt]: Agree! The moment of the decision whether prism is more-validation or more-mocking is coming :)

File: packages/http/package.json (hunk 10 above)
Comment (XVincentX) [reply packages/http/package.json]: Saturday task: create typings for such library :trollface:

File: test-harness/specs/parameters-ac10.oas3.txt [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+                properties: {}
+====server====
+mock -p 4010 ${document}
+====command====
+curl -i http://127.0.0.1:4010/path//foo
+====expect====
+HTTP/1.1 422 Unprocessable Entity
+content-type: application/problem+json
+
+{"type":"https://stoplight.io/prism/errors#UNPROCESSABLE_ENTITY","title":"Invalid request body payload","status":422,"detail":"Your request body is not valid and no HTTP validation response was found in the spec, so Prism is generating this error for you.","validation":[{"location":["path"],"severity":"Error","code":"required","message":"should have required property 'param'"}]}
```
Comment (lag-of-death) [inline test-harness/specs/parameters-ac10.oas3.txt]:
```suggestion
{"type":"https://stoplight.io/prism/errors#UNPROCESSABLE_ENTITY","title":"Invalid request body payload","status":422,"detail":"Your request is not valid and no HTTP validation response was found in the spec, so Prism is generating this error for you.","validation":[{"location":["path"],"severity":"Error","code":"required","message":"should have required property 'param'"}]}
```

Not sure if this is a change for this PR, but maybe? Surely saying `request body` here ain't right

File: test-harness/specs/parameters-ac9.oas3.txt [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+                properties: {}
+====server====
+mock -p 4010 ${document}
+====command====
+curl -i http://127.0.0.1:4010/path/fp
+====expect====
+HTTP/1.1 422 Unprocessable Entity
+content-type: application/problem+json
+
+{"type":"https://stoplight.io/prism/errors#UNPROCESSABLE_ENTITY","title":"Invalid request body payload","status":422,"detail":"Your request body is not valid and no HTTP validation response was found in the spec, so Prism is generating this error for you.","validation":[{"location":["path","param"],"severity":"Error","code":"pattern","message":"should match pattern \"^[A-Z]+$\""}]}
```
Comment (lag-of-death) [inline test-harness/specs/parameters-ac9.oas3.txt]: ditto

File: test-harness/specs/parameters-ac10.oas3.txt (hunk 13 above)
Comment (StefanDywersant) [reply test-harness/specs/parameters-ac10.oas3.txt]: Looks like not right, I'll check where is that coming from. Good spot!
